stan/math/rev/mat/functor/algebra_solver_fp.hpp

```diff
@@ -0,0 +1,342 @@
+   * @param f_tol Function tolenrance
+   * @param max_num_steps max nb. of iterations.
+   */
+  template <typename T1>
+  Eigen::Matrix<stan::math::var, -1, 1> solve(
+      const Eigen::Matrix<T1, -1, 1>& x,
+      const Eigen::Matrix<stan::math::var, -1, 1>& y,
+      KinsolFixedPointEnv<F>& env, double f_tol, int max_num_steps) {
+    using stan::math::value_of;
+    using stan::math::var;
+
+    // FP solution
+    Eigen::VectorXd xd(value_of(x));
+    kinsol_solve_fp(xd, env, f_tol, max_num_steps);
+
+    fp_jac_type jac_sol;
+    return jac_sol(xd, y, env);
+  }
+};
+
+/**
+ * Return a fixed pointer to the specified system of algebraic
+ * equations of form
+ *
+ * x = F(x; theta)
+ *
+ * given an initial guess x, and parameters theta and data. Use the
+ * KINSOL solver from the SUNDIALS suite.
+ *
+ * The user can also specify the scaling controls, the function
+ * tolerance, and the maximum number of steps.
+ *
+ * @tparam F type of equation system function.
+ * @tparam T type of initial guess vector.
+ * @param[in] f Functor that evaluated the system of equations.
+ * @param[in] x Vector of starting values.
+ * @param[in] y Parameter vector for the equation system. The function
+ *            is overloaded to treat y as a vector of doubles or of a
+ *            a template type T.
+ * @param[in] x_r Continuous data vector for the equation system.
+ * @param[in] x_i Integer data vector for the equation system.
+ * @param[in, out] msgs The print stream for warning messages.
+ * @param[in] u_scale scaling vector for solution
+ * @param[in] f_scale scaling vector for residual
+ * @param[in] f_tol determines whether roots are acceptable.
+ * @param[in] max_num_steps maximum number of function evaluations.
+ *  * @throw <code>std::invalid_argument</code> if x has size zero.
+ * @throw <code>std::invalid_argument</code> if x has non-finite elements.
+ * @throw <code>std::invalid_argument</code> if y has non-finite elements.
+ * @throw <code>std::invalid_argument</code> if dat has non-finite elements.
+ * @throw <code>std::invalid_argument</code> if dat_int has non-finite elements.
+ * @throw <code>std::invalid_argument</code> if scaled_step_size is strictly
+ * negative.
+ * @throw <code>std::invalid_argument</code> if function_tolerance is strictly
+ * negative.
+ * @throw <code>std::invalid_argument</code> if max_num_steps is not positive.
+ * @throw <code>boost::math::evaluation_error</code> (which is a subclass of
+ * <code>std::runtime_error</code>) if solver exceeds max_num_steps.
+ */
+template <typename F, typename T1, typename T2>
+Eigen::Matrix<T2, -1, 1> algebra_solver_fp(
+    const F& f, const Eigen::Matrix<T1, -1, 1>& x,
+    const Eigen::Matrix<T2, -1, 1>& y, const std::vector<double>& x_r,
+    const std::vector<int>& x_i, const std::vector<double>& u_scale,
+    const std::vector<double>& f_scale, std::ostream* msgs = nullptr,
+    double f_tol = 1e-8,
+    int max_num_steps = 200) {  // NOLINT(runtime/int)
```

Comment:
They are specified just like `x_r`, `x_i`, etc. I find reason to believe there's a default setting that can be applied in general.

Reply:
Let me clarify my question: how does a user choose which values to give to `u_scale` and `f_scale`? Since these values have a fixed type, does that mean they won't depend on the parameters? It seems the scale of the solution could depend on the model parameters. All that to say, finding `u_scale` and `f_scale` may be advanced for the standard user.

KINSOL provides defaults, why not use them and make the argument optional?

> I find reason to believe there's a default setting that can be applied in general.

I don't understand what you mean: are you saying there defaults that can be applied, or that having defaults would be a bad idea?

Reply:
I meant to say "I find no reason", sorry for the typo.

> All that to say, finding u_scale and f_scale may be advanced for the standard user.

It doesn't mean we shouldn't guide them through correct way of using math libs. It's user's responsibility to make sure his math checks out. To answer your question of what if scaling is param dependent, we can allow these two scaling vectors `var` type so they can be calculated on the fly, but this doesn't make user's life easier, as it's not supposed to be.

ODE solver is advanced for the standard user, many people would argue. 
> KINSOL provides defaults, why not use them and make the argument optional?

Does it? Could you point me to the documentation?

Reply:
> we can allow these two scaling vectors var type so they can be calculated on the fly

Yes, this would be good. When I first implemented `algebra_solve()`, I forced the initial guess to be data only and this caused some issues.

I also agree that this kind of method is designed for more advanced users, but our benchmark should be people who run `algebra_solve`. That said, user can resort to `algebra_solve_*` if they prefer.

> Does it? Could you point me to the documentation?

You're right, it might not.

In the section on Anderson's acceleration, the manual states D_F (which is built from `f_scale`) can be the identity or a scaling matrix. I propose to offer the identity (or in this case a vector of 1's) as a naive default. This is what I do in `kinsol_solve` -- with the caveat the user currently cannot control this scaling vector if they want. There are a few problems where this will be the right choice. 

We then need to clearly state in the documentation that the user should be prepared to modify this if needed. Another option would be to set the scaling matrix based on the initial guess (though we run into issues if the user passes a vector of 0's as the initial guess).

Reply:
> I propose to offer the identity (or in this case a vector of 1's) as a naive default. This is what I do in kinsol_solve (with the caveat the user currently cannot control these scaling vectors). There are a few problems where this naive choice will be the right one.

But there are more problems that non-scaling won't be the right one, as isotropy is much less common than anisotropy in this world. Keep in mind that `u_scale`/`f_scale` is not something that KINSOL is comfortable with to provide a default, so there has to be a very good reason for us to venture into assumptions of use cases. The concept of convergence is always tied to "in what sense", so asking user to provide scaling is for him to understand this.

Reply:
Ok, your argument is sound.

I still think a default would be fine, but on the other hand, it's straightforward for the user to pass in a vector of 1's if they choose to. And you're right, better to make them think about it than to have them ignore it. We'll have to think about this when we write the "pedantic" version of the Newton solver.

Reply:
> We'll have to think about this when we write the "pedantic" version of the Newton solver.

yeah, had that in mind when reviewing the PR. Also the constraints will be a good addon.